lib/RelatedResourceProviders/GroupFoldersRelatedResourceProvider.php

```diff
@@ -0,0 +1,191 @@
+	private array $folders = [];
+
+	public function __construct(
+		private IRootFolder $rootFolder,
+		private IURLGenerator $urlGenerator,
+		private IL10N $l10n,
+		private FilesShareRequest $filesShareRequest,
+	) {
+		try {
+			$this->circlesManager = Server::get(CirclesManager::class);
+			$this->folderManager = Server::get(FolderManager::class);
+			$this->folders = $this->folderManager->getAllFolders();
+		} catch (ContainerExceptionInterface|AutoloadNotAllowedException $e) {
+		}
+	}
+
+	public function getProviderId(): string {
+		return self::PROVIDER_ID;
+	}
+
+	public function loadWeightCalculator(): array {
+		return [];
+	}
+
+	public function getRelatedFromItem(string $itemId): ?IRelatedResource {
+		if ($this->circlesManager === null) {
+			return null;
+		}
+
+		$itemId = (int)$itemId;
+		if ($itemId < 1) {
+			return null;
+		}
+
+		// need to get the groupfolders parent, based on itemId
+		try {
+			$folder = $this->getFolder($itemId);
+		} catch (GroupFolderNotFoundException $e) {
+			return null;
+		}
+
+		$related = $this->convertToRelatedResource($folder);
+		$this->processApplicableMap($related, $folder['groups'] ?? []);
+
+		return $related;
+	}
+
+	public function getItemsAvailableToEntity(FederatedUser $entity): array {
+		$items = [];
+		foreach ($this->folders as $folder) {
+			foreach ($folder['groups'] as $k => $entry) {
+				if ($entity->getBasedOn()->getSource() === Member::TYPE_GROUP
+					&& $entry['type'] === 'group'
+					&& $k === $entity->getUserId()) {
+					$items[] = (string)$folder['id'];
+				}
+
+				if ($entity->getBasedOn()->getSource() === Member::TYPE_CIRCLE
+					&& $entry['type'] === 'circle'
+					&& $k === $entity->getSingleId()) {
+					$items[] = (string)$folder['id'];
+				}
+			}
+		}
+
+		return $items;
+	}
+
+
+	public function improveRelatedResource(IRelatedResource $entry): void {
+	}
+
+
+	public function convertToRelatedResource(array $folderData): IRelatedResource {
+		$related = new RelatedResource(self::PROVIDER_ID, (string)($folderData['id'] ?? 0));
+		$folderName = $folderData['mount_point'] ?? 'groupfolder';
+		$related->setTitle($folderName);
+		$related->setSubtitle($this->l10n->t('Group Folder'));
+		$related->setTooltip($this->l10n->t('Group Folder "%s"', '/' . $folderName . '/'));
+
+		$related->setIcon(
+			$this->urlGenerator->getAbsoluteURL(
+				$this->urlGenerator->imagePath('groupfolders', 'app-dark.svg')
+			)
+		);
+
+		$related->setUrl(
+			$this->urlGenerator->linkToRouteAbsolute(
+				'files.view.index',
+				['dir' => '/' . $folderName]
+			)
+		);
+
+		$related->setMetaArray(RelatedResource::ITEM_KEYWORDS, [$folderName]);
+
+		return $related;
+	}
+
+	/**
+	 * @param RelatedResource $related
+	 * @param array $applicableMap
+	 */
+	public function processApplicableMap(RelatedResource $related, array $applicableMap): void {
+		foreach ($applicableMap as $k => $entry) {
+			$entityId = '';
+			if ($entry['type'] === 'circle') {
+				$entityId = $k;
```

Comment:
```suggestion
				$entityId = (string)$k;
```
You need to cast here because PHP does weird shit with array keys, and numeric string will come out as integers.